### PR TITLE
Additions to persistence PR

### PR DESCRIPTION
**Pull request by BenjaminBossan**
- Slight change in signature: `save(obj, file)` instead of
  `save(file, obj)`, as it's more consistent with existing saving functions
- Rename `cls` to `cls_` in `get_instance_method` to prevent variable name
  reuse
- Add `Persister` class that allows users to add their custom methods for
  getting and loading objects.
- Add support for `Pipeline` and `FeatureUnion`
- Add support for basic numpy functions
- Add test scaffold + a couple of tests (some of them `xfail`ing for now)

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
The discussed changes @adrinjalali

File: skops/_persist.py [hunk 1]
```diff
@@ -1,24 +1,37 @@
...
+from typing import Any
 from uuid import uuid4
 from zipfile import ZipFile
 
 import numpy as np
+import sklearn
 from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+
+import skops
```
Comment (adrinjalali) [inline skops/_persist.py]: why do we need this?

File: skops/_persist.py [hunk 2]
```diff
@@ -1,24 +1,37 @@
...
 
 def gettype(state):
     if "__module__" in state and "__class__" in state:
-        return getattr(importlib.import_module(state["__module__"]), state["__class__"])
+        return _import_obj(state["__module__"], state["__class__"])
     return None
 
 
-def BaseEstimator_get_state(obj, dst):
+def BaseEstimator_get_state(obj, dst, methods):
```
Comment (adrinjalali) [inline skops/_persist.py]: I would hold on with adding the extra method dispatch thing and think about that in a later PR, there are other ways to do it as well I think.

File: skops/_persist.py [hunk 3]
```diff
@@ -88,73 +139,192 @@ def dict_get_state(obj, dst):
...
+    persister = Persister(protocol=protocol)
+    return persister.load(file)
 
 
-def load(file):
-    input_zip = ZipFile(file)
-    schema = input_zip.read("schema.json")
-    return get_instance_method(json.loads(schema))(json.loads(schema), input_zip)
-    # return {name: input_zip.read(name) for name in input_zip.namelist()}
+class Persister:
```
Comment (adrinjalali) [inline skops/_persist.py]: if we hold on for the extra methods thing here, then we also don't need this class for now.

File: skops/test_persist.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        assert type(val0) == type(val1)
+        if isinstance(val0, BaseEstimator):
+            assert_params_equal(val0.get_params(), val1.get_params())
+        elif isinstance(val0, (np.ndarray, np.generic)):
+            np.testing.assert_allclose(val0, val1)
+        else:
+            assert val0 == val1
+
+
+class EstimatorBaseTester:
```
Comment (adrinjalali) [inline skops/test_persist.py]: For common tests, I rather do the way I now put in the PR: https://github.com/skops-dev/skops/pull/128/files#diff-9ce8f30521c2b9fb6a1b2f0813e522810362fa16cc9a8be1660c4f5637dbcce1

It reduces the boilerplate code I think.

File: skops/test_persist.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        return union.fit(X, y)
+
+    def test_get_output(self, estimator, loaded, data):
+        X, _ = data
+        X_original = estimator.transform(X)
+        X_loaded = loaded.transform(X)
+        np.testing.assert_allclose(X_original, X_loaded)
+
+
+class TestClfPipelineAndFeatureUnion(EstimatorBaseTester):
```
Comment (adrinjalali) [inline skops/test_persist.py]: happy to keep this one though, for a more in depth test here.

**Review by adrinjalali — COMMENTED** [review]
Thanks @BenjaminBossan , this is quite a bit of work already done!

**Comment by adrinjalali** [comment]
I also put the test under `skops/tests` instead of `skops/`

**Comment by BenjaminBossan** [comment]
@adrinjalali I made the requested changes. Some comments:

I can't seem to be able to fetch your latest changes, so after my merge all the tests you added will break since I changed the signature of `save`.

As for the testing approach, I was thinking about using the existing sklearn utilities to add tests but decided against it because:

1. They rely on private functions.
2. Most of the tested objects still fail - is there an easy way to mark a subset of them to be expected to fail? I don't know any, hence I wanted to opt them in explicitly.
3. Some special cases are not tested there, e.g. the deeply nested pipeline/feature union, `FunctionTransformer` with specific function types, etc., though that should not be too hard to add.

It's still okay to go that way but we need a solution for 2.

**Comment by adrinjalali** [comment]
Yes, I agree we should do the more complicated tests as well, that's what I meant by this one: https://github.com/adrinjalali/skops/pull/1#discussion_r964611730

We'd have added tests specific to certain classes.

As for the failing tests, this is what we do as it's usually done in sklearn, with explicit ignore lists: https://github.com/scikit-learn/scikit-learn/blob/670133dbc42ebd9f79552984316bc2fcfd208e2e/sklearn/tests/test_docstrings.py#L14

As for using private methods, I agree, but rather still use them and work upstream to make them public or a part of the developer API. We can make sure things work for future releases by adding a test against the nightly builds.

**Comment by adrinjalali** [comment]
This looks good, I'll merge and fix the rest on the PR.

**Comment by BenjaminBossan** [comment]
> As for the failing tests, this is what we do as it's usually done in sklearn, with explicit ignore lists: https://github.com/scikit-learn/scikit-learn/blob/670133dbc42ebd9f79552984316bc2fcfd208e2e/sklearn/tests/test_docstrings.py#L14

I think for now we would have an easier time with opting in instead of opting out :D

**Pull request merged**